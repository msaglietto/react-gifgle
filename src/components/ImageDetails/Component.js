import React from 'react'
import PropTypes from 'prop-types'

import './style.css'

class Component extends React.Component {
  static displayName = 'ImageDetails'

  static propTypes = {
    getImage: PropTypes.func.isRequired,
    image: PropTypes.object.isRequired,
    match: PropTypes.object.isRequired,
    isLoading: PropTypes.bool.isRequired,
    error: PropTypes.string.isRequired
  }

  componentWillMount () {
    const { slug } = this.props.match.params
    this.props.getImage(slug)
  }

  componentWillReceiveProps (nextProps) {
    const params = this.props.match
    const { slug } = nextProps.match
    if (params.slug !== slug) {
      this.props.getImage(slug)
    }
  }

  render () {
    const { image, isLoading, error } = this.props

    if (isLoading) {
      return 'Loading ...'
    }

    if (error) {
      return 'Sorry it has been an error while fetching the image'
    }

    return (
      <div className='ImageDetails'>
        <h2>{image.title}</h2>
        <img src={image.original} alt={image.title} />
      </div>
    )
  }
}

export default Component
