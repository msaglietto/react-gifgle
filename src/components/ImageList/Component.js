import React from 'react'
import PropTypes from 'prop-types'

const propTypes = {
  images: PropTypes.array
}

const defaultProps = {
  images: []
}

export const Component = props => {
  const imagesComp = props.images.map((image) => <img key={image.id} src={image.images.fixed_height_small.url} alt={image.title} />)

  return (
    <div className='ImageList'>
      {imagesComp}
    </div>
  )
}

Component.propTypes = propTypes
Component.defaultProps = defaultProps
Component.displayName = 'ImageList'

export default Component
