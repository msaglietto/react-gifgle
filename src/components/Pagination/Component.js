import React from 'react'
import PropTypes from 'prop-types'

import './style.css'

const propTypes = {
  currentPage: PropTypes.number,
  perPage: PropTypes.number,
  totalItems: PropTypes.number.isRequired,
  onPageClick: PropTypes.func.isRequired,
  pagesToShow: PropTypes.number
}

const defaultProps = {
  currentPage: 0,
  perPage: 25,
  pagesToShow: 10
}

const handePageClick = (props, page) => () => props.onPageClick(page)

const button = (props, page, label) => (
  <li key={label}>
    <button onClick={handePageClick(props, page)}> {label} </button>
  </li>
)

const Component = props => {
  if (!props.totalItems) {
    return null
  }

  const maxPages = Math.ceil(props.totalItems / props.perPage)
  const range = Math.min(maxPages, props.pagesToShow)
  const midPoint = props.currentPage - Math.floor(props.pagesToShow / 2)
  const showFromPage = Math.max(
    0,
    Math.min(maxPages - props.pagesToShow, Math.max(0, midPoint))
  )

  return (
    <ul className='Pagination'>
      {button(props, 0, '<<')}
      {button(props, Math.max(props.currentPage - 1, 0), '<')}
      {Array.apply(null, Array(range)).map((_, i) => {
        const atPage = i + showFromPage
        return (atPage === props.currentPage
          ? <li key={atPage + 1}>{atPage + 1}</li>
          : button(props, atPage, atPage + 1)
        )
      })}
      {button(props, Math.min(props.currentPage + 1, maxPages - 1), '>')}
      {button(props, maxPages - 1, '>>')}
    </ul>
  )
}

Component.propTypes = propTypes
Component.defaultProps = defaultProps
Component.displayName = 'Pagination'

export default Component
