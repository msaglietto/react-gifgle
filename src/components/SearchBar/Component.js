import React from 'react'
import PropTypes from 'prop-types'

const handleInputChange = props => event => {
  props.onInputChange(event.target.value)
}

const propTypes = {
  onInputChange: PropTypes.func.isRequired
}

const defaultProps = {

}

export const Component = props => (
  <input placeholder='Enter gif topic' onChange={handleInputChange(props)} />
)

Component.propTypes = propTypes
Component.defaultProps = defaultProps
Component.displayName = 'SearchBar'

export default Component
