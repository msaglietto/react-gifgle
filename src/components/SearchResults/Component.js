import React from 'react'
import PropTypes from 'prop-types'

import ImageList from '../ImageList'
import Pagination from '../Pagination'

class Component extends React.Component {
  static displayName = 'SearchResults'

  static propTypes = {
    getImages: PropTypes.func.isRequired,
    images: PropTypes.array.isRequired,
    page: PropTypes.number.isRequired,
    total: PropTypes.number.isRequired,
    match: PropTypes.object.isRequired,
    isLoading: PropTypes.bool.isRequired,
    error: PropTypes.string.isRequired
  }

  componentWillMount () {
    const { query, page } = this.props.match.params
    this.props.getImages(query, parseInt(page || 0, 10))
  }

  componentWillReceiveProps (nexProps) {
    const { params } = this.props.match
    const { query, page } = nexProps.match.params
    if (query !== params.query || page !== params.page) {
      this.props.getImages(query, parseInt(page, 10))
    }
  }

  handlePageChange = (page) => {
    const { query } = this.props.match.params
    this.props.history.push(`/search/${query}/${page}`)
  }

  handleImageClick = (image) => {
    this.props.history.push(`/show/${image.slug}`)
  }

  render () {
    const { isLoading, error } = this.props

    if (isLoading) {
      return 'Loading ...'
    }

    if (error) {
      return 'Sorry it has been an error while fetching the image'
    }

    return (
      <section className='SearchReults section'>
        <div class='message is-info' style={{'marginTop': 10}}>
          <h2 className='message-body'>You are seeing the results of {this.props.match.params.query || 'trending'} Gifts</h2>
        </div>

        <ImageList images={this.props.images} onImageClick={this.handleImageClick} />

        <div className='SearchResults-pagination' style={{'marginTop': 10}}>
          <Pagination
            currentPage={this.props.page}
            totalItems={this.props.total}
            onPageClick={this.handlePageChange}
          />
        </div>
      </section>
    )
  }
}

export default Component
