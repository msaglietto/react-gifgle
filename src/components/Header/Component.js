import React from 'react'
import PropTypes from 'prop-types'
import { Link } from 'react-router-dom'

import SearchBar from '../SearchBar'

import logo from './logo.svg'
import './style.css'

const contextTypes = {
  router: PropTypes.object
}

const handleNewSearch = (context) => (query) => context.router.history.push(`/search/${query}`)

export const Component = (props, context) => (
  <header className='Header'>
    <Link to='/'><img src={logo} className='Header-logo' alt='GIFGLE Logo' /></Link>
    <Link to='/'><h1 className='Header-title'>GIFGLE</h1></Link>
    <div className='Header-SearchBar'>
      <SearchBar onNewSearch={handleNewSearch(context)} />
    </div>
  </header>
)

Component.displayName = 'Header'
Component.contextTypes = contextTypes

export default Component
