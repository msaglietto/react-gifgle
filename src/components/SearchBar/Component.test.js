/* eslint-env jest */
import React from 'react'
import { shallow } from 'enzyme'

import { Component } from './Component'

describe('SearchBar', () => {
  it('should render equal to the snapshot', () => {
    const props = {
      onInputChange: () => {}
    }
    const wrapper = shallow(<Component {...props} />)

    expect(wrapper).toMatchSnapshot()
  })

  it('should call prop on change', () => {
    const props = {
      onInputChange: jest.fn()
    }
    const wrapper = shallow(<Component {...props} />)
    wrapper.find('input').simulate('change', {target: {value: 'test'}})

    expect(props.onInputChange).toHaveBeenCalledWith('test')
  })
})
