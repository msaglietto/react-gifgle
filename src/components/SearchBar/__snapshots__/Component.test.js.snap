// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchBar should render equal to the snapshot 1`] = `
<input
  onChange={[Function]}
  placeholder="Enter gif topic"
/>
`;
